pkg/proxy/endpointslicecache.go

```diff
@@ -82,56 +105,92 @@ func standardEndpointInfo(ep *BaseEndpointInfo) Endpoint {
 	return ep
 }
 
-// Update a slice in the cache.
-func (cache *EndpointSliceCache) Update(endpointSlice *discovery.EndpointSlice) {
+// updatePending updates a pending slice in the cache.
+func (cache *EndpointSliceCache) updatePending(endpointSlice *discovery.EndpointSlice, remove bool) bool {
 	serviceKey, sliceKey, err := endpointSliceCacheKeys(endpointSlice)
 	if err != nil {
 		klog.Warningf("Error getting endpoint slice cache keys: %v", err)
-		return
+		return false
 	}
 
 	esInfo := &endpointSliceInfo{
 		Ports:     endpointSlice.Ports,
 		Endpoints: []*endpointInfo{},
+		Remove:    remove,
 	}
-	for _, endpoint := range endpointSlice.Endpoints {
-		if endpoint.Conditions.Ready == nil || *endpoint.Conditions.Ready == true {
-			esInfo.Endpoints = append(esInfo.Endpoints, &endpointInfo{
-				Addresses: endpoint.Addresses,
-				Topology:  endpoint.Topology,
-			})
+
+	if !remove {
+		for _, endpoint := range endpointSlice.Endpoints {
+			if endpoint.Conditions.Ready == nil || *endpoint.Conditions.Ready == true {
+				esInfo.Endpoints = append(esInfo.Endpoints, &endpointInfo{
+					Addresses: endpoint.Addresses,
+					Topology:  endpoint.Topology,
+				})
+			}
 		}
+
+		sort.Sort(byAddress(esInfo.Endpoints))
 	}
-	if _, exists := cache.sliceByServiceMap[serviceKey]; !exists {
-		cache.sliceByServiceMap[serviceKey] = map[string]*endpointSliceInfo{}
+
+	cache.lock.Lock()
+	defer cache.lock.Unlock()
+
+	if _, ok := cache.trackerByServiceMap[serviceKey]; !ok {
+		cache.trackerByServiceMap[serviceKey] = newEndpointSliceTracker()
 	}
-	cache.sliceByServiceMap[serviceKey][sliceKey] = esInfo
+
+	changed := cache.esInfoChanged(serviceKey, sliceKey, esInfo)
+
+	if changed {
+		cache.trackerByServiceMap[serviceKey].pending[sliceKey] = esInfo
+	}
+
+	return changed
```

Comment:
It may helped the performance more by not comparing the difference and just trigger sync as always. 

But it may not make any difference given the test result.

Reply:
Unfortunately both the `EndpointChangesPending` metric and the last change trigger times rely on this being available on EndpointSlice addition (computing at proxier sync would not be helpful). The profiling I've done suggests that the full `updatePending` function, including this diffing, takes 0.27% of total compute time, so I think it's worth maintaining the metrics as they exist.